File: app/src/main/java/com/krystianwsul/checkme/domainmodel/BackendNotifier.java
package com.krystianwsul.checkme.domainmodel;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.TextUtils;
import android.util.Log;

import com.android.volley.NoConnectionError;
import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.TimeoutError;
import com.android.volley.VolleyError;
import com.android.volley.toolbox.StringRequest;
import com.android.volley.toolbox.Volley;
import com.annimon.stream.Collectors;
import com.annimon.stream.Stream;
import com.krystianwsul.checkme.MyCrashlytics;
import com.krystianwsul.checkme.firebase.DatabaseWrapper;
import com.krystianwsul.checkme.firebase.RemoteProject;

import junit.framework.Assert;

import java.util.Collection;
import java.util.List;
import java.util.Set;

class BackendNotifier {
    private static final String PREFIX = "http://check-me-add47.appspot.com/notify?";

    @NonNull
    static String getUrl(@NonNull Set<String> projects, boolean production, @NonNull Collection<String> userKeys, @Nullable String senderToken) {
        Assert.assertTrue(!projects.isEmpty());

        List<String> parameters = Stream.of(projects)
                .map(project -> "projects=" + project)
                .collect(Collectors.toList());

        parameters.addAll(Stream.of(userKeys)
                .map(userKey -> "userKeys=" + userKey)
                .collect(Collectors.toList()));

        if (production)
            parameters.add("production=1");

        parameters.add("senderToken=" + senderToken);

        return PREFIX + TextUtils.join("&", parameters);
    }

    BackendNotifier(@NonNull Context context, @NonNull Set<RemoteProject> remoteProjects, @NonNull UserInfo userInfo, @NonNull Collection<String> userKeys) {
        String root = DatabaseWrapper.getRoot();

        boolean production;
        switch (root) {
            case "development":
                production = false;
                break;
            case "production":
                production = true;
                break;
            default:
                throw new IllegalArgumentException();
        }

        Set<String> projectIds = Stream.of(remoteProjects)
                .map(RemoteProject::getId)
                .collect(Collectors.toSet());

        String url = getUrl(projectIds, production, userKeys, userInfo.mToken);
        Assert.assertTrue(!TextUtils.isEmpty(url));

        run(context, url);
    }

    private void run(@NonNull Context context, @NonNull String url) {
        Assert.assertTrue(!TextUtils.isEmpty(url));

        RequestQueue queue = Volley.newRequestQueue(context.getApplicationContext());

        StringRequest stringRequest = new StringRequest(
                Request.Method.GET, url,
                (response) -> Log.e("asdf", "BackendNotifier response:" + response),
                (VolleyError error) -> {
                    if ((error instanceof TimeoutError) || (error instanceof NoConnectionError)) {
                        Log.e("asdf", "BackendNotifier error", error);
                        return;
                    }

                    MyCrashlytics.logException(error);
                });

        queue.add(stringRequest);
    }
}

File: app/src/main/java/com/krystianwsul/checkme/domainmodel/BackendNotifier.kt
package com.krystianwsul.checkme.domainmodel

import android.util.Log
import com.android.volley.NoConnectionError
import com.android.volley.Request
import com.android.volley.ServerError
import com.android.volley.TimeoutError
import com.android.volley.toolbox.StringRequest
import com.android.volley.toolbox.Volley
import com.krystianwsul.checkme.MyApplication
import com.krystianwsul.checkme.MyCrashlytics
import com.krystianwsul.checkme.firebase.AndroidDatabaseWrapper
import com.krystianwsul.common.domain.DeviceInfo
import com.krystianwsul.common.firebase.models.project.Project
import com.krystianwsul.common.utils.ProjectKey
import com.krystianwsul.common.utils.UserKey

object BackendNotifier {

    private const val PREFIX = "https://us-central1-check-me-add47.cloudfunctions.net/notify?"

    private fun getUrl(projects: Set<ProjectKey<*>>, production: Boolean, userKeys: Collection<UserKey>, senderToken: String): String {
        check(projects.isNotEmpty() || !userKeys.isEmpty())

        val parameters = projects.map { "projects=${it.key}" }.toMutableSet()

        parameters.addAll(userKeys.map { "userKeys=${it.key}" })

        if (production)
            parameters.add("production=1")

        parameters.add("senderToken=$senderToken")

        return PREFIX + parameters.joinToString("&")
    }

    fun notify(projects: Set<Project<*>>, deviceInfo: DeviceInfo, userKeys: Collection<UserKey>) {
        val production = when (AndroidDatabaseWrapper.root) {
            "development" -> false
            "production" -> true
            else -> throw IllegalArgumentException()
        }

        val projectIds = projects.map { it.projectKey }.toSet()

        val url = getUrl(projectIds, production, userKeys, deviceInfo.token!!)
        check(url.isNotEmpty())

        run(url)
    }

    private fun run(url: String) {
        check(url.isNotEmpty())

        val queue = Volley.newRequestQueue(MyApplication.instance)
        Log.e("asdf", "BackendNotifier url: $url")
        val stringRequest = StringRequest(
                Request.Method.GET,
                url,
                {
                    Log.e("asdf", "BackendNotifier response:")

                    val lines = it.lines()

                    lines.forEach { Log.e("asdf", it) }

                    lines.firstOrNull { it.startsWith("error") }?.let {
                        val error = (lines.indexOf(it) until lines.size).joinToString("\n") { lines[it] }

                        MyCrashlytics.logException(BackendException(error))
                    }
                },
                {
                    when (it) {
                        is TimeoutError, is NoConnectionError -> Log.e("asdf", "BackendNotifier error", it)
                        is ServerError -> MyCrashlytics.logException(ServerErrorWrapper("message: ${it.message},  networkResponse: " + it.networkResponse?.let { "code: ${it.statusCode}, data: ${String(it.data)}" }))
                        else -> MyCrashlytics.logException(it)
                    }
                })

        queue.add(stringRequest)
    }

    private class BackendException(message: String) : Exception(message)

    private class ServerErrorWrapper(message: String) : Exception(message)
}
